src/Rules/Register.hs

```diff
@@ -1,44 +1,97 @@
-module Rules.Register (registerPackage) where
+module Rules.Register (registerPackages) where
 
 import Base
 import Context
 import GHC
+import Settings
 import Target
 import Utilities
 
--- TODO: Simplify.
+import Distribution.ParseUtils
+import qualified Distribution.Compat.ReadP as Parse
+import Distribution.Version (Version)
+
+import Hadrian.Expression
+import Hadrian.Haskell.Cabal.Parse as Cabal
+
+parseCabalName :: String -> Maybe (String, Version)
+parseCabalName = readPToMaybe parse
+  where parse = (,) <$> (parsePackageName <* Parse.char '-') <*> parseOptVersion
+
 -- | Build rules for registering packages and initialising package databases
 -- by running the @ghc-pkg@ utility.
-registerPackage :: [(Resource, Int)] -> Context -> Rules ()
-registerPackage rs context@Context {..} = do
-    when (stage == Stage0) $ do
-        -- Packages @ghc-boot@ and @ghc-boot-th@ both match the @ghc-boot*@
-        -- pattern, therefore we need to use priorities to match the right rule.
-        -- TODO: Get rid of this hack.
-        "//" ++ stage0PackageDbDir -/- pkgName package ++ "*.conf" %%>
-            buildConf rs context
+registerPackages :: [(Resource, Int)] -> Context -> Rules ()
```

Comment:
Why is it renamed to `registerPackages` if it still accepts a single `Context`? This rule still registers just a single package.

Reply:
This is a generic rule now. Instead of generating a rule for each package, we have one for any package. It extracts the package name from the path, and updates the Context with the relevant package as needed.

Reply:
@angerman I'm confused: if this rule now works for any package, why does it take `Context` as an argument? Presumably the stage can also be extracted from the build path.

Is `Context` here used only to communicate the build way?

Reply:
If we drop context here, we'll have to adapt `copyConf` and `buildConf` as well, which in turn will drag in all other called functions that somehow want to pass a context.

We could--I guess pass in Way--and try to parse the `Stage` from the file pattern to recreate the `Context`. I'm not certain that is much better than the current situation?

Reply:
Well, at the moment the type signature is rather confusing. The function name tells us that this is a generic rule, but we are passing in a full `Context`, apparently whose `package` field is unused.

If parsing `Stage` is a pain, let's simply pass in both `Stage` and `Way` -- it's much more direct than passing in `Context`.